Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,835 @@
+                    slice.moveWriterIndex(to: slice.writerIndex - Int(padding))
+                }
+                self.state = .idle
+                return HTTP2Frame(streamID: streamID, flags: flags, payload: .data(.byteBuffer(slice)))
+            }
+        case .accumulatingHeaderBlockFragments(let header, var buffer):
+            // special case since we only accept CONTINUATION frames in this state
+            guard let continuationHeader = bytes.readFrameHeader() else {
+                return nil      // don't have another header yet
+            }
+            
+            guard continuationHeader.type == 0x9 /* CONTINUATION */ else {
+                // invalid frame: this is a connection error
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            
+            let flags = HTTP2Frame.FrameFlags(rawValue: continuationHeader.flags)
+            
+            var cbytes: ByteBuffer
+            if bytes.readableBytes >= continuationHeader.length {
+                cbytes = bytes.readSlice(length: continuationHeader.length)!
+                buffer.write(buffer: &cbytes)
+                
+                if (flags.contains(.endHeaders)) {
+                    // we're done: send out a frame
+                    return try self.emitSyntheticHeadersFrame(header, bytes: &buffer)
+                } else {
+                    self.state = .accumulatingHeaderBlockFragments(header, buffer)
+                }
+            } else {
+                cbytes = bytes
+                bytes.clear()
+                self.state = .accumulatingContinuationPayload(originalHeader: header, headerBytes: buffer, continuationHeader: continuationHeader, continuationBytes: cbytes)
+            }
+            return nil  // no frame to return yet
+            
+        case .accumulatingContinuationPayload(let originalHeader, var headerPayload, let continuationHeader, var continuationBytes):
+            let required = continuationHeader.length - continuationBytes.readableBytes
+            if bytes.readableBytes >= required {
+                var slice = bytes.readSlice(length: required)!
+                headerPayload.write(buffer: &continuationBytes)
+                headerPayload.write(buffer: &slice)
+                
+                // pretend it's one big HEADERS or PUSH_PROMISE frame
+                return try self.emitSyntheticHeadersFrame(originalHeader, bytes: &headerPayload)
+            } else {
+                continuationBytes.write(buffer: &bytes)
+                return nil
+            }
+        }
+    }
+    
+    private mutating func emitSyntheticHeadersFrame(_ originalHeader: FrameHeader, bytes: inout ByteBuffer) throws -> HTTP2Frame? {
+        let newHeader = FrameHeader(length: bytes.readableBytes, type: originalHeader.type,
+                                    flags: originalHeader.flags | HTTP2Frame.FrameFlags.endHeaders.rawValue,
```

Comment:
This looks wrong to me: specifically, I think it mishandles PADDING on HEADERS frames. It keeps the PADDING flag intact, but does not preserve either the PADDING length or the PADDING bytes, which is almost certainly going to cause a HPACK error.

I recommend writing a test for this too before fixing it. :smile:

Reply:
I manually eat the padding and remove that flag in the new state machine.